traefik crashes when kubernetes secret is used for basic auth

<!--
DO NOT FILE ISSUES FOR GENERAL SUPPORT QUESTIONS.

The issue tracker is for reporting bugs and feature requests only.
For end-user related support questions, refer to one of the following:

- Stack Overflow (using the "traefik" tag): https://stackoverflow.com/questions/tagged/traefik
- the Traefik community Slack channel: https://traefik.herokuapp.com

-->


### Do you want to request a *feature* or report a *bug*?
BUG.
<!--
If you intend to ask a support question: DO NOT FILE AN ISSUE.
-->

### What did you do?
Applied #1707 to get access to secrets
Installed deployment of latest traefik image with 3 replicas.
Deployed k8s-dashboard and made ingress rule with basic auth + secret.
<!--

HOW TO WRITE A GOOD ISSUE?

- Respect the issue template as more as possible.
- If it's possible use the command `traefik bug`. See https://www.youtube.com/watch?v=Lyz62L8m93I. 
- The title must be short and descriptive.
- Explain the conditions which led you to write this issue: the context.
- The context should lead to something, an idea or a problem that you’re facing.
- Remain clear and concise.
- Format your messages to help the reader focus on what matters and understand the structure of your message, use Markdown syntax https://help.github.com/articles/github-flavored-markdown

-->

### What did you expect to see?

All replicas to be working. Not only 2 of them.

### What did you see instead?

2\3 are up and the third one couldn't get existing secret.


### Output of `traefik version`: (_What version of Traefik are you using?_)
```
Version:      v1.3.1
Codename:     raclette
Go version:   go1.8.3
Built:        2017-06-16_11:21:48AM
OS/Arch:      linux/amd64
```
### What is your environment & configuration (arguments, toml, provider, platform, ...)?

```toml
debug = true
logLevel = "INFO"
defaultEntryPoints = ["http","https"]

[entryPoints]
	[entryPoints.http]
	address = ":80"
	[entryPoints.https]
	address = ":443"

[web]
address = ":8080"
	[web.auth.basic]
	users = [ "test:$apr1$Y8X0T234$8hoaJeKIka4wJi13yETbK." ]

[kubernetes]
namespaces = [ "default", "kube-system" ]
```
<!--
Add more configuration information here.
-->


### If applicable, please paste the log output in debug mode (`--debug` switch)

```
time="2017-06-27T15:32:58Z" level=info msg="Traefik version v1.3.1 built on 2017-06-16_11:21:48AM" 
time="2017-06-27T15:32:58Z" level=info msg="Using TOML configuration file /config/traefik.toml" 
time="2017-06-27T15:32:58Z" level=debug msg="Global configuration loaded {"GraceTimeOut":10000000000,"Debug":true,"CheckNewVersion":true,"AccessLogsFile":"","TraefikLogsFile":"","LogLevel":"DEBUG","EntryPoints":{"http":{"Network":"","Address":":80","TLS":null,"Redirect":null,"Auth":null,"Compress":false},"https":{"Network":"","Address":":443","TLS":null,"Redirect":null,"Auth":null,"Compress":false}},"Cluster":null,"Constraints":[],"ACME":null,"DefaultEntryPoints":["http","https"],"ProvidersThrottleDuration":2000000000,"MaxIdleConnsPerHost":200,"IdleTimeout":180000000000,"InsecureSkipVerify":false,"Retry":null,"HealthCheck":{"Interval":30000000000},"Docker":null,"File":null,"Web":{"Address":":8080","CertFile":"","KeyFile":"","ReadOnly":false,"Statistics":null,"Metrics":null,"Path":"","Auth":{"Basic":{"Users":["test:$apr1$Y8X0T234$8hoaJeKIka4wJi13yETbK."],"UsersFile":""},"Digest":null,"HeaderField":""}},"Marathon":null,"Consul":null,"ConsulCatalog":null,"Etcd":null,"Zookeeper":null,"Boltdb":null,"Kubernetes":{"Watch":true,"Filename":"","Constraints":[],"Endpoint":"","Token":"","CertAuthFilePath":"","DisablePassHostHeaders":false,"Namespaces":["default","kube-system"],"LabelSelector":""},"Mesos":null,"Eureka":null,"ECS":null,"Rancher":null,"DynamoDB":null}" 
time="2017-06-27T15:32:58Z" level=info msg="Preparing server https &{Network: Address::443 TLS:<nil> Redirect:<nil> Auth:<nil> Compress:false}" 
time="2017-06-27T15:32:58Z" level=info msg="Preparing server http &{Network: Address::80 TLS:<nil> Redirect:<nil> Auth:<nil> Compress:false}" 
time="2017-06-27T15:32:58Z" level=info msg="Starting provider *server.WebProvider {"Address":":8080","CertFile":"","KeyFile":"","ReadOnly":false,"Statistics":null,"Metrics":null,"Path":"","Auth":{"Basic":{"Users":["test:$apr1$Y8X0T234$8hoaJeKIka4wJi13yETbK."],"UsersFile":""},"Digest":null,"HeaderField":""}}" 
time="2017-06-27T15:32:58Z" level=info msg="Starting provider *kubernetes.Provider {"Watch":true,"Filename":"","Constraints":[],"Endpoint":"","Token":"","CertAuthFilePath":"","DisablePassHostHeaders":false,"Namespaces":["default","kube-system"],"LabelSelector":""}" 
time="2017-06-27T15:32:58Z" level=info msg="Starting server on :80" 
time="2017-06-27T15:32:58Z" level=info msg="Starting server on :443" 
time="2017-06-27T15:32:58Z" level=info msg="Creating in-cluster Provider client
" 
time="2017-06-27T15:32:58Z" level=debug msg="Using label selector: ''" 
time="2017-06-27T15:32:59Z" level=debug msg="Received event from kubernetes &Endpoints{ObjectMeta:ObjectMeta{Name:tiller-deploy,GenerateName:,Namespace:kube-system,SelfLink:/api/v1/namespaces/kube-system/endpoints/tiller-deploy,UID:db0c7d32-1b98-11e7-8d60-0025905d1042,ResourceVersion:9558297,Generation:0,CreationTimestamp:2017-04-07 13:48:18 +0000 UTC,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{app: helm,name: tiller,},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,},Subsets:[{[{192.168.0.4  0xc4207bcbc0 ObjectReference{Kind:Pod,Namespace:kube-system,Name:tiller-deploy-1332042700-pb18c,UID:daebcdb3-1b98-11e7-8d60-0025905d1042,APIVersion:,ResourceVersion:9558296,FieldPath:,}}] [] [{tiller 44134 TCP}]}],}" 
time="2017-06-27T15:32:59Z" level=error msg="Provider connection error secret "kube-system"/"k8s-dashboard-basic-auth" not found, retrying in 651.281427ms" 
ERROR: logging before flag.Parse: E0627 15:32:59.062442       1 runtime.go:66] Observed a panic: "send on closed channel" (send on closed channel)
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:72
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:65
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:514
/usr/local/go/src/runtime/panic.go:489
/usr/local/go/src/runtime/chan.go:180
/usr/local/go/src/runtime/chan.go:604
/go/src/github.com/containous/traefik/provider/kubernetes/client.go:136
/go/src/github.com/containous/traefik/provider/kubernetes/client.go:143
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:173
<autogenerated>:50
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:251
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/delta_fifo.go:420
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:131
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:102
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait/wait.go:96
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait/wait.go:97
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait/wait.go:52
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:102
/usr/local/go/src/runtime/asm_amd64.s:2197
ERROR: logging before flag.Parse: E0627 15:32:59.062615       1 runtime.go:66] Observed a panic: "send on closed channel" (send on closed channel)
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:72
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:65
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:514
/usr/local/go/src/runtime/panic.go:489
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:58
/usr/local/go/src/runtime/asm_amd64.s:514
/usr/local/go/src/runtime/panic.go:489
/usr/local/go/src/runtime/chan.go:180
/usr/local/go/src/runtime/chan.go:604
/go/src/github.com/containous/traefik/provider/kubernetes/client.go:136
/go/src/github.com/containous/traefik/provider/kubernetes/client.go:143
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:173
<autogenerated>:50
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:251
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/delta_fifo.go:420
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:131
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:102
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait/wait.go:96
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait/wait.go:97
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait/wait.go:52
/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:102
/usr/local/go/src/runtime/asm_amd64.s:2197
panic: send on closed channel [recovered]
	panic: send on closed channel [recovered]
	panic: send on closed channel

goroutine 39 [running]:
github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:58 +0x126
panic(0x1ca6940, 0xc4208ee540)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/runtime/runtime.go:58 +0x126
panic(0x1ca6940, 0xc4208ee540)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/containous/traefik/provider/kubernetes.eventHandlerFunc(0xc420112d20, 0x1f5e6e0, 0xc4200f3800)
	/go/src/github.com/containous/traefik/provider/kubernetes/client.go:136 +0x55
github.com/containous/traefik/provider/kubernetes.newResourceEventHandlerFuncs.func1(0x1f5e6e0, 0xc4200f3800)
	/go/src/github.com/containous/traefik/provider/kubernetes/client.go:143 +0x3e
github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnAdd(0xc42032dc50, 0xc42032dc60, 0xc42032dc70, 0x1f5e6e0, 0xc4200f3800)
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:173 +0x49
github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache.(*ResourceEventHandlerFuncs).OnAdd(0xc4207ce300, 0x1f5e6e0, 0xc4200f3800)
	<autogenerated>:50 +0x73
github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache.NewInformer.func1(0x1cd5ac0, 0xc4208ff2e0, 0x1cd5ac0, 0xc4208ff2e0)
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:251 +0x24d
github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache.(*DeltaFIFO).Pop(0xc4206602c0, 0xc4205bf320, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/delta_fifo.go:420 +0x218
github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache.(*Controller).processLoop(0xc42015f110)
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:131 +0x3c
github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache.(*Controller).(github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache.processLoop)-fm()
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:102 +0x2a
github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait.JitterUntil.func1(0xc4207d87b0)
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait/wait.go:96 +0x5e
github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait.JitterUntil(0xc4207a7fb0, 0x3b9aca00, 0x0, 0x1b63001, 0xc420112c60)
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait/wait.go:97 +0xad
github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait.Until(0xc4207d87b0, 0x3b9aca00, 0xc420112c60)
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/pkg/util/wait/wait.go:52 +0x4d
github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache.(*Controller).Run(0xc42015f110, 0xc420112c60)
	/go/src/github.com/containous/traefik/vendor/k8s.io/client-go/tools/cache/controller.go:102 +0x1c7
created by github.com/containous/traefik/provider/kubernetes.(*clientImpl).WatchSecrets
	/go/src/github.com/containous/traefik/provider/kubernetes/client.go:227 +0x28e
```

--- a/provider/kubernetes/client.go
+++ b/provider/kubernetes/client.go
@@ -263,7 +263,7 @@ func (c *clientImpl) WatchAll(labelSelector string, stopCh <-chan struct{}) (<-c
 // fireEvent checks if all controllers have synced before firing
 // Used after startup or a reconnect
 func (c *clientImpl) fireEvent(event interface{}, eventCh chan interface{}) {
-	if !c.ingController.HasSynced() || !c.svcController.HasSynced() || !c.epController.HasSynced() {
+	if !c.ingController.HasSynced() || !c.svcController.HasSynced() || !c.epController.HasSynced() || !c.secController.HasSynced() {
 		return
 	}
 	eventHandlerFunc(eventCh, event)

--- a/provider/kubernetes/kubernetes.go
+++ b/provider/kubernetes/kubernetes.go
@@ -110,11 +110,11 @@ func (p *Provider) Provide(configurationChan chan<- types.ConfigMessage, pool *s
 		}
 
 		notify := func(err error, time time.Duration) {
-			log.Errorf("Provider connection error %+v, retrying in %s", err, time)
+			log.Errorf("Provider connection error: %s; retrying in %s", err, time)
 		}
 		err := backoff.RetryNotify(safe.OperationWithRecover(operation), job.NewBackOff(backoff.NewExponentialBackOff()), notify)
 		if err != nil {
-			log.Errorf("Cannot connect to Provider server %+v", err)
+			log.Errorf("Cannot connect to Provider: %s", err)
 		}
 	})
 
@@ -290,18 +290,15 @@ func handleBasicAuthConfig(i *v1beta1.Ingress, k8sClient Client) ([]string, erro
 		return nil, nil
 	}
 	if strings.ToLower(authType) != "basic" {
-		return nil, fmt.Errorf("unsupported auth-type: %q", authType)
+		return nil, fmt.Errorf("unsupported auth-type on annotation ingress.kubernetes.io/auth-type: %q", authType)
 	}
 	authSecret := i.Annotations["ingress.kubernetes.io/auth-secret"]
 	if authSecret == "" {
-		return nil, errors.New("auth-secret annotation must be set")
+		return nil, errors.New("auth-secret annotation ingress.kubernetes.io/auth-secret must be set")
 	}
 	basicAuthCreds, err := loadAuthCredentials(i.Namespace, authSecret, k8sClient)
 	if err != nil {
-		return nil, err
-	}
-	if len(basicAuthCreds) == 0 {
-		return nil, errors.New("secret file without credentials")
+		return nil, fmt.Errorf("failed to load auth credentials: %s", err)
 	}
 	return basicAuthCreds, nil
 }
@@ -314,9 +311,9 @@ func loadAuthCredentials(namespace, secretName string, k8sClient Client) ([]stri
 	case !ok:
 		return nil, fmt.Errorf("secret %q/%q not found", namespace, secretName)
 	case secret == nil:
-		return nil, errors.New("secret data must not be nil")
+		return nil, fmt.Errorf("data for secret %q/%q must not be nil", namespace, secretName)
 	case len(secret.Data) != 1:
-		return nil, errors.New("secret must contain single element only")
+		return nil, fmt.Errorf("found %d elements for secret %q/%q, must be single element exactly", len(secret.Data), namespace, secretName)
 	default:
 	}
 	var firstSecret []byte
@@ -331,6 +328,10 @@ func loadAuthCredentials(namespace, secretName string, k8sClient Client) ([]stri
 			creds = append(creds, cred)
 		}
 	}
+	if len(creds) == 0 {
+		return nil, fmt.Errorf("secret %q/%q does not contain any credentials", namespace, secretName)
+	}
+
 	return creds, nil
 }
 

--- a/provider/kubernetes/kubernetes.go
+++ b/provider/kubernetes/kubernetes.go
@@ -171,7 +171,8 @@ func (p *Provider) loadIngresses(k8sClient Client) (*types.Configuration, error)
 				if _, exists := templateObjects.Frontends[r.Host+pa.Path]; !exists {
 					basicAuthCreds, err := handleBasicAuthConfig(i, k8sClient)
 					if err != nil {
-						return nil, err
+						log.Errorf("Failed to retrieve basic auth configuration for ingress %s/%s: %s", i.ObjectMeta.Namespace, i.ObjectMeta.Name, err)
+						continue
 					}
 					templateObjects.Frontends[r.Host+pa.Path] = &types.Frontend{
 						Backend:        r.Host + pa.Path,

